### ISSUE TITLE
Record Completion for Unions

### ISSUE THREAD

**Issue by ari-becker** [dhall-lang/dhall-lang#821]
I went ahead and brought [`dhall-kops`](https://github.com/coralogix/dhall-kops) up-to-date for v11 of the language standard. The new record completion semantics are an absolutely tremendous addition to the language and are an incredible improvement in language ergonomics.

They do, however, have a hole when it comes to unions. Consider ['HookSpec'](https://github.com/coralogix/dhall-kops/blob/v0.4.0/api/v1alpha2/HookSpec.dhall):

```dhall
let ExecContainerAction = ./ExecContainerAction.dhall

let DockerImage =
      { Type =
          { disabled : Optional Bool
          , roles : Optional (List Text)
          , requires : Optional (List Text)
          , before : Optional (List Text)
          , execContainer : ExecContainerAction.Type
          }
      , default =
          { disabled = None Bool
          , roles = None (List Text)
          , requires = None (List Text)
          , before = None (List Text)
          }
      }

let SystemdUnit =
      { Type =
          { name : Text
          , disabled : Optional Bool
          , roles : Optional (List Text)
          , requires : Optional (List Text)
          , before : Optional (List Text)
          , manifest : Text
          , useRawManifest : Optional Bool
          }
      , default =
          { disabled = None Bool
          , roles = None (List Text)
          , requires = None (List Text)
          , before = None (List Text)
          , useRawManifest = None Bool
          }
      }

let Union = < DockerImage : DockerImage.Type | SystemdUnit : SystemdUnit.Type >

in  { DockerImage = DockerImage, SystemdUnit = SystemdUnit, Union = Union }
```

Because the record completion semantics are `T::r => T.default // r : T.Type`, and the `//` operator is undefined for unions, I still have to write (as seen in the [tests](https://github.com/coralogix/dhall-kops/blob/v0.4.0/tests/api/v1alpha2/tests.dhall)):

```dhall
Kops.HookSpec.Union.DockerImage
  Kops.HookSpec.DockerImage::{
    , execContainer = Kops.ExecContainerAction::{ image = "example" }
    }
```

which seems, in light of the ergonomics improvements afforded by the record completion operator... suboptimal. Ideally, I could just write:

```dhall
Kops.HookSpec.DockerImage::{
, execContainer = Kops.ExecContainerAction::{ image = "example" }
}
```

, where `Kops.HookSpec` is the union (instead of a record with field `Union` which holds the union), and the union itself would be defined as (suggested idea):

```dhall
< DockerImage :: DockerImage
| SystemdUnit :: SystemdUnit
>
```

, for `DockerImage` and `SystemdUnit` as originally defined (i.e. a record with both `Type` and `default`).

Relates to #175 (and probably others, didn't look too hard).

**Comment by sjakobi** [comment]
This looks like it could be useful for the use case described in https://github.com/dhall-lang/dhall-lang/issues/809.

Does the following description accurately describe your idea?

Given a "schema-union" ("union-schema"?)

```dhall
let SchemaUnion = < X :: { Type = X, default = x0 } | Y :: { Type = Y, default = y0 } >
```

you'd like

```dhall
SchemaUnion.X::x1
```

to construct the value

```dhall
< X : X | Y : Y >.X (x0 // x1)
```

**Comment by ari-becker** [comment]
@sjakobi yep, seems right. It would probably require that, in a union in which there is a single alternative using `::` syntax, then all alternatives must use `::` syntax, excepting empty alternatives which would imply `< Empty :: { Type = {} , default = {=} } >`.

**Comment by sjakobi** [comment]
Is the syntactic convenience of writing

```dhall
Kops.HookSpec.DockerImage::{
, execContainer = Kops.ExecContainerAction::{ image = "example" }
}
```

instead of

```dhall
Kops.HookSpec.Union.DockerImage
  Kops.HookSpec.DockerImage::{
    , execContainer = Kops.ExecContainerAction::{ image = "example" }
    }
```

the only benefit of this proposal?

IMHO it would be a bit disproportionate add an entire new language construct for this.

**Comment by ari-becker** [comment]
I mean, was the syntactic convenience of writing `T::r` the only benefit of supporting record completion before? I think ergonomics is important.

If you want more bang (at the cost of a little more buck) I can imagine something where, in addition to `Type` and `default`, arbitrary functions could be written, e.g. a `render` function for `T -> Text`, and then if each alternative defines that function then there could be some desugaring into a `merge` with handles block. But I think that's a little much for now.

**Comment by sjakobi** [comment]
> I mean, was the syntactic convenience of writing `T::r` the only benefit of supporting record completion before? I think ergonomics is important.

So far, my impression is that record completion was a fairly big jump in terms of expressiveness and syntactic convenience. In the example above it seems that the main benefit is that we can avoid the explicit application of the union constructor – this effect is linear in the number of union value constructions, but apparently with a pretty small factor.

On the implementation side, record completion is very cheap: it can be desugared to existing language constructs. OTOH these "schema-unions" seem to require a rather far-reaching extensions of the syntax and type system!

I wonder whether we can get around extending the type system for these "schema-unions" though.
Instead of writing

```dhall
< DockerImage :: DockerImage
| SystemdUnit :: SystemdUnit
>
```

we could (ab)use records like this:

```dhall
let Kops.HookSpec =
      { DockerImage = DockerImage
      , SystemdUnit = SystemdUnit
      }
```

and then use a keyword to construct the desired union value, e.g.

```dhall
complete
  Kops.HookSpec
  DockerImage
  { execContainer = Kops.ExecContainerAction::{ image = "example" } }
```

Or some kind of ternary operator?

```dhall
Kops.HookSpec<:DockerImage:>{
  , execContainer = Kops.ExecContainerAction::{ image = "example" }
  }
```

**Comment by ari-becker** [comment]
From an ergonomics standpoint, I don't really like either of those options, because it seems like it would be easy to lose sight of what the desugared union looks like, and thus make it more difficult for users to debug in case something goes wrong. The principal thing I like about the `< X :: { Type = X , default = x } | Y :: { Type = Y , default = y } >` is that it cleanly mentally maps (at least in my mind) to `< X : X | Y : Y >`, especially for users familiar with the `{ Type = X , default = x }` record completion model.

Thinking a little further, I think this is coming from a place of feeling like it's a little too difficult to work with records that are wrapped by unions, and maybe this is more of a request to focus on union ergonomics in general.
